### PR TITLE
Add region/cut selection tool to the slice viewer

### LINKED ISSUE #27467: Sliceviewer: slicing with thickness
This is axis aligned only

Use binmd/cutmd to create a finite thickness 2-d slice of the data

Use binmd/cutmd to create a finite thickness 1-d slice of the data

### LINKED ISSUE #28966: Sliceviewer cannot view some saved reduced workspaces
**Original reporter:** ISIS/Spencer

### Expected behavior

Sliceviewer opens when viewing all MatrixWorkspace instances.

### Actual behavior
For some reduced MSG datasets the sliceviewer catches an unknown exception and just logs a warning that it cannot open:

```
Could not open slice viewer for workspace 'iris26176_graphite002_red'.
Exception: unknown
```

### Steps to reproduce the behavior
* Load `Testing/Data/DocTest/irs26176_graphite002_red.nxs`
* Try to open the slice viewer

### Platforms affected
All

### LINKED ISSUE #29005: Crash Plotting after Sliceviewer
This was found by new students in the ISIS Mantid team during the Mantid Basic Course

### Expected behavior
Sliceviewer and normal plots close cleanly and there is no crash!
### Actual behavior
It is possible to open and close the sliceviewer, try to plot a few spectra and then Mantid hard crashes
### Steps to reproduce the behavior
These steps do not work every time. If they do not crash Mantid, then go through the steps a second time!

- Load MAR11060 from TrainingCourseData
- Load EMU00020884.nxs from TrainingCourseData
- Open the sliceviewer on the MAR11060 workspace
- CLOSE sliceviewer
- Plot spectra 1-10 tiled of the EMU00020884 workspace

### Platforms affected
Mac and Windows latest nightly/master 5.0.20200714.1059

Mac Python crash report: [crashMBC.txt](https://github.com/mantidproject/mantid/files/4924577/crashMBC.txt)

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Adds a tool allowing a region to be selected in the slice viewer that causes the lineplots to display the integrated signal over the orthogonal dimension rather than that from a single pixel.

Keyboard shortcuts have been added to enable extracting roi/cuts to separate workspaces in a similar manner to the instrument view.

**To test:**

Tests need to be done on both MD workspaces and Matrix workspaces. Generate a fake MD workspace with

```
mdws = CreateMDWorkspace(Dimensions=3,
                         Extents='-10,10,-10,10,-10,10',
                         Names='A,B,C',
                         Units='r.l.u.,r.l.u.,r.l.u.',
                         Frames='HKL,HKL,HKL')
expt_info = CreateSampleWorkspace()
mdws.addExperimentInfo(expt_info)
SetUB(mdws, 1, 1, 1, 90, 90, 90)
FakeMDEventData(InputWorkspace=mdws, PeakParams='500000,-1,0,0,3')
FakeMDEventData(InputWorkspace=mdws, PeakParams='500000,5,0,0,3')
```

Generate a test MatrixWorkspace, the file is in the unit test data, with:

```
filename = 'ZOOM00006113.nxs'
LoadISISNexus(Filename=filename, OutputWorkspace=filename)
GroupDetectors(InputWorkspace=filename, OutputWorkspace=filename+'_88_to_92',
               GroupingPattern='45056+45568+46080+46592+, 45057+45569+46081+46593+, 45058+45570+46082+46594+, 45059+45571+46083+46595+, 45060+45572+46084+46596+, 45061+45573+46085+46597+, 45062+45574+46086+46598+, 45063+45575+46087+46599+, 45064+45576+46088+46600+, 45065+45577+46089+46601+, 45066+45578+46090+46602+, 45067+45579+46091+46603+, 45068+45580+46092+46604+, 45069+45581+46093+46605+, 45070+45582+46094+46606+, 45071+45583+46095+46607+, 45072+45584+46096+46608+, 45073+45585+46097+46609+, 45074+45586+46098+46610+, 45075+45587+46099+46611+, 45076+45588+46100+46612+, 45077+45589+46101+46613+, 45078+45590+46102+46614+, 45079+45591+46103+46615+, 45080+45592+46104+46616+, 45081+45593+46105+46617+, 45082+45594+46106+46618+, 45083+45595+46107+46619+, 45084+45596+46108+46620+, 45085+45597+46109+46621+, 45086+45598+46110+46622+, 45087+45599+46111+46623+, 45088+45600+46112+46624+, 45089+45601+46113+46625+, 45090+45602+46114+46626+, 45091+45603+46115+46627+, 45092+45604+46116+46628+, 45093+45605+46117+46629+, 45094+45606+46118+46630+, 45095+45607+46119+46631+, 45096+45608+46120+46632+, 45097+45609+46121+46633+, 45098+45610+46122+46634+, 45099+45611+46123+46635+, 45100+45612+46124+46636+, 45101+45613+46125+46637+, 45102+45614+46126+46638+, 45103+45615+46127+46639+, 45104+45616+46128+46640+, 45105+45617+46129+46641+, 45106+45618+46130+46642+, 45107+45619+46131+46643+, 45108+45620+46132+46644+, 45109+45621+46133+46645+, 45110+45622+46134+46646+, 45111+45623+46135+46647+, 45112+45624+46136+46648+, 45113+45625+46137+46649+, 45114+45626+46138+46650+, 45115+45627+46139+46651+, 45116+45628+46140+46652+, 45117+45629+46141+46653+, 45118+45630+46142+46654+, 45119+45631+46143+46655+, 45120+45632+46144+46656+, 45121+45633+46145+46657+, 45122+45634+46146+46658+, 45123+45635+46147+46659+, 45124+45636+46148+46660+, 45125+45637+46149+46661+, 45126+45638+46150+46662+, 45127+45639+46151+46663+, 45128+45640+46152+46664+, 45129+45641+46153+46665+, 45130+45642+46154+46666+, 45131+45643+46155+46667+, 45132+45644+46156+46668+, 45133+45645+46157+46669+, 45134+45646+46158+46670+, 45135+45647+46159+46671+, 45136+45648+46160+46672+, 45137+45649+46161+46673+, 45138+45650+46162+46674+, 45139+45651+46163+46675+, 45140+45652+46164+46676+, 45141+45653+46165+46677+, 45142+45654+46166+46678+, 45143+45655+46167+46679+, 45144+45656+46168+46680+, 45145+45657+46169+46681+, 45146+45658+46170+46682+, 45147+45659+46171+46683+, 45148+45660+46172+46684+, 45149+45661+46173+46685+, 45150+45662+46174+46686+, 45151+45663+46175+46687+, 45152+45664+46176+46688+, 45153+45665+46177+46689+, 45154+45666+46178+46690+, 45155+45667+46179+46691+, 45156+45668+46180+46692+, 45157+45669+46181+46693+, 45158+45670+46182+46694+, 45159+45671+46183+46695+, 45160+45672+46184+46696+, 45161+45673+46185+46697+, 45162+45674+46186+46698+, 45163+45675+46187+46699+, 45164+45676+46188+46700+, 45165+45677+46189+46701+, 45166+45678+46190+46702+, 45167+45679+46191+46703+, 45168+45680+46192+46704+, 45169+45681+46193+46705+, 45170+45682+46194+46706+, 45171+45683+46195+46707+, 45172+45684+46196+46708+, 45173+45685+46197+46709+, 45174+45686+46198+46710+, 45175+45687+46199+46711+, 45176+45688+46200+46712+, 45177+45689+46201+46713+, 45178+45690+46202+46714+, 45179+45691+46203+46715+, 45180+45692+46204+46716+, 45181+45693+46205+46717+, 45182+45694+46206+46718+, 45183+45695+46207+46719+, 45184+45696+46208+46720+, 45185+45697+46209+46721+, 45186+45698+46210+46722+, 45187+45699+46211+46723+, 45188+45700+46212+46724+, 45189+45701+46213+46725+, 45190+45702+46214+46726+, 45191+45703+46215+46727+, 45192+45704+46216+46728+, 45193+45705+46217+46729+, 45194+45706+46218+46730+, 45195+45707+46219+46731+, 45196+45708+46220+46732+, 45197+45709+46221+46733+, 45198+45710+46222+46734+, 45199+45711+46223+46735+, 45200+45712+46224+46736+, 45201+45713+46225+46737+, 45202+45714+46226+46738+, 45203+45715+46227+46739+, 45204+45716+46228+46740+, 45205+45717+46229+46741+, 45206+45718+46230+46742+, 45207+45719+46231+46743+, 45208+45720+46232+46744+, 45209+45721+46233+46745+, 45210+45722+46234+46746+, 45211+45723+46235+46747+, 45212+45724+46236+46748+, 45213+45725+46237+46749+, 45214+45726+46238+46750+, 45215+45727+46239+46751+, 45216+45728+46240+46752+, 45217+45729+46241+46753+, 45218+45730+46242+46754+, 45219+45731+46243+46755+, 45220+45732+46244+46756+, 45221+45733+46245+46757+, 45222+45734+46246+46758+, 45223+45735+46247+46759+, 45224+45736+46248+46760+, 45225+45737+46249+46761+, 45226+45738+46250+46762+, 45227+45739+46251+46763+, 45228+45740+46252+46764+, 45229+45741+46253+46765+, 45230+45742+46254+46766+, 45231+45743+46255+46767+, 45232+45744+46256+46768+, 45233+45745+46257+46769+, 45234+45746+46258+46770+, 45235+45747+46259+46771+, 45236+45748+46260+46772+, 45237+45749+46261+46773+, 45238+45750+46262+46774+, 45239+45751+46263+46775+, 45240+45752+46264+46776+, 45241+45753+46265+46777+, 45242+45754+46266+46778+, 45243+45755+46267+46779+, 45244+45756+46268+46780+, 45245+45757+46269+46781+, 45246+45758+46270+46782+, 45247+45759+46271+46783+, 45248+45760+46272+46784+, 45249+45761+46273+46785+, 45250+45762+46274+46786+, 45251+45763+46275+46787+, 45252+45764+46276+46788+, 45253+45765+46277+46789+, 45254+45766+46278+46790+, 45255+45767+46279+46791+, 45256+45768+46280+46792+, 45257+45769+46281+46793+, 45258+45770+46282+46794+, 45259+45771+46283+46795+, 45260+45772+46284+46796+, 45261+45773+46285+46797+, 45262+45774+46286+46798+, 45263+45775+46287+46799+, 45264+45776+46288+46800+, 45265+45777+46289+46801+, 45266+45778+46290+46802+, 45267+45779+46291+46803+, 45268+45780+46292+46804+, 45269+45781+46293+46805+, 45270+45782+46294+46806+, 45271+45783+46295+46807+, 45272+45784+46296+46808+, 45273+45785+46297+46809+, 45274+45786+46298+46810+, 45275+45787+46299+46811+, 45276+45788+46300+46812+, 45277+45789+46301+46813+, 45278+45790+46302+46814+, 45279+45791+46303+46815+, 45280+45792+46304+46816+, 45281+45793+46305+46817+, 45282+45794+46306+46818+, 45283+45795+46307+46819+, 45284+45796+46308+46820+, 45285+45797+46309+46821+, 45286+45798+46310+46822+, 45287+45799+46311+46823+, 45288+45800+46312+46824+, 45289+45801+46313+46825+, 45290+45802+46314+46826+, 45291+45803+46315+46827+, 45292+45804+46316+46828+, 45293+45805+46317+46829+, 45294+45806+46318+46830+, 45295+45807+46319+46831+, 45296+45808+46320+46832+, 45297+45809+46321+46833+, 45298+45810+46322+46834+, 45299+45811+46323+46835+, 45300+45812+46324+46836+, 45301+45813+46325+46837+, 45302+45814+46326+46838+, 45303+45815+46327+46839+, 45304+45816+46328+46840+, 45305+45817+46329+46841+, 45306+45818+46330+46842+, 45307+45819+46331+46843+, 45308+45820+46332+46844+, 45309+45821+46333+46845+, 45310+45822+46334+46846+, 45311+45823+46335+46847+, 45312+45824+46336+46848+, 45313+45825+46337+46849+, 45314+45826+46338+46850+, 45315+45827+46339+46851+, 45316+45828+46340+46852+, 45317+45829+46341+46853+, 45318+45830+46342+46854+, 45319+45831+46343+46855+, 45320+45832+46344+46856+, 45321+45833+46345+46857+, 45322+45834+46346+46858+, 45323+45835+46347+46859+, 45324+45836+46348+46860+, 45325+45837+46349+46861+, 45326+45838+46350+46862+, 45327+45839+46351+46863+, 45328+45840+46352+46864+, 45329+45841+46353+46865+, 45330+45842+46354+46866+, 45331+45843+46355+46867+, 45332+45844+46356+46868+, 45333+45845+46357+46869+, 45334+45846+46358+46870+, 45335+45847+46359+46871+, 45336+45848+46360+46872+, 45337+45849+46361+46873+, 45338+45850+46362+46874+, 45339+45851+46363+46875+, 45340+45852+46364+46876+, 45341+45853+46365+46877+, 45342+45854+46366+46878+, 45343+45855+46367+46879+, 45344+45856+46368+46880+, 45345+45857+46369+46881+, 45346+45858+46370+46882+, 45347+45859+46371+46883+, 45348+45860+46372+46884+, 45349+45861+46373+46885+, 45350+45862+46374+46886+, 45351+45863+46375+46887+, 45352+45864+46376+46888+, 45353+45865+46377+46889+, 45354+45866+46378+46890+, 45355+45867+46379+46891+, 45356+45868+46380+46892+, 45357+45869+46381+46893+, 45358+45870+46382+46894+, 45359+45871+46383+46895+, 45360+45872+46384+46896+, 45361+45873+46385+46897+, 45362+45874+46386+46898+, 45363+45875+46387+46899+, 45364+45876+46388+46900+, 45365+45877+46389+46901+, 45366+45878+46390+46902+, 45367+45879+46391+46903+, 45368+45880+46392+46904+, 45369+45881+46393+46905+, 45370+45882+46394+46906+, 45371+45883+46395+46907+, 45372+45884+46396+46908+, 45373+45885+46397+46909+, 45374+45886+46398+46910+, 45375+45887+46399+46911+, 45376+45888+46400+46912+, 45377+45889+46401+46913+, 45378+45890+46402+46914+, 45379+45891+46403+46915+, 45380+45892+46404+46916+, 45381+45893+46405+46917+, 45382+45894+46406+46918+, 45383+45895+46407+46919+, 45384+45896+46408+46920+, 45385+45897+46409+46921+, 45386+45898+46410+46922+, 45387+45899+46411+46923+, 45388+45900+46412+46924+, 45389+45901+46413+46925+, 45390+45902+46414+46926+, 45391+45903+46415+46927+, 45392+45904+46416+46928+, 45393+45905+46417+46929+, 45394+45906+46418+46930+, 45395+45907+46419+46931+, 45396+45908+46420+46932+, 45397+45909+46421+46933+, 45398+45910+46422+46934+, 45399+45911+46423+46935+, 45400+45912+46424+46936+, 45401+45913+46425+46937+, 45402+45914+46426+46938+, 45403+45915+46427+46939+, 45404+45916+46428+46940+, 45405+45917+46429+46941+, 45406+45918+46430+46942+, 45407+45919+46431+46943+, 45408+45920+46432+46944+, 45409+45921+46433+46945+, 45410+45922+46434+46946+, 45411+45923+46435+46947+, 45412+45924+46436+46948+, 45413+45925+46437+46949+, 45414+45926+46438+46950+, 45415+45927+46439+46951+, 45416+45928+46440+46952+, 45417+45929+46441+46953+, 45418+45930+46442+46954+, 45419+45931+46443+46955+, 45420+45932+46444+46956+, 45421+45933+46445+46957+, 45422+45934+46446+46958+, 45423+45935+46447+46959+, 45424+45936+46448+46960+, 45425+45937+46449+46961+, 45426+45938+46450+46962+, 45427+45939+46451+46963+, 45428+45940+46452+46964+, 45429+45941+46453+46965+, 45430+45942+46454+46966+, 45431+45943+46455+46967+, 45432+45944+46456+46968+, 45433+45945+46457+46969+, 45434+45946+46458+46970+, 45435+45947+46459+46971+, 45436+45948+46460+46972+, 45437+45949+46461+46973+, 45438+45950+46462+46974+, 45439+45951+46463+46975+, 45440+45952+46464+46976+, 45441+45953+46465+46977+, 45442+45954+46466+46978+, 45443+45955+46467+46979+, 45444+45956+46468+46980+, 45445+45957+46469+46981+, 45446+45958+46470+46982+, 45447+45959+46471+46983+, 45448+45960+46472+46984+, 45449+45961+46473+46985+, 45450+45962+46474+46986+, 45451+45963+46475+46987+, 45452+45964+46476+46988+, 45453+45965+46477+46989+, 45454+45966+46478+46990+, 45455+45967+46479+46991+, 45456+45968+46480+46992+, 45457+45969+46481+46993+, 45458+45970+46482+46994+, 45459+45971+46483+46995+, 45460+45972+46484+46996+, 45461+45973+46485+46997+, 45462+45974+46486+46998+, 45463+45975+46487+46999+, 45464+45976+46488+47000+, 45465+45977+46489+47001+, 45466+45978+46490+47002+, 45467+45979+46491+47003+, 45468+45980+46492+47004+, 45469+45981+46493+47005+, 45470+45982+46494+47006+, 45471+45983+46495+47007+, 45472+45984+46496+47008+, 45473+45985+46497+47009+, 45474+45986+46498+47010+, 45475+45987+46499+47011+, 45476+45988+46500+47012+, 45477+45989+46501+47013+, 45478+45990+46502+47014+, 45479+45991+46503+47015+, 45480+45992+46504+47016+, 45481+45993+46505+47017+, 45482+45994+46506+47018+, 45483+45995+46507+47019+, 45484+45996+46508+47020+, 45485+45997+46509+47021+, 45486+45998+46510+47022+, 45487+45999+46511+47023+, 45488+46000+46512+47024+, 45489+46001+46513+47025+, 45490+46002+46514+47026+, 45491+46003+46515+47027+, 45492+46004+46516+47028+, 45493+46005+46517+47029+, 45494+46006+46518+47030+, 45495+46007+46519+47031+, 45496+46008+46520+47032+, 45497+46009+46521+47033+, 45498+46010+46522+47034+, 45499+46011+46523+47035+, 45500+46012+46524+47036+, 45501+46013+46525+47037+, 45502+46014+46526+47038+, 45503+46015+46527+47039+, 45504+46016+46528+47040+, 45505+46017+46529+47041+, 45506+46018+46530+47042+, 45507+46019+46531+47043+, 45508+46020+46532+47044+, 45509+46021+46533+47045+, 45510+46022+46534+47046+, 45511+46023+46535+47047+, 45512+46024+46536+47048+, 45513+46025+46537+47049+, 45514+46026+46538+47050+, 45515+46027+46539+47051+, 45516+46028+46540+47052+, 45517+46029+46541+47053+, 45518+46030+46542+47054+, 45519+46031+46543+47055+, 45520+46032+46544+47056+, 45521+46033+46545+47057+, 45522+46034+46546+47058+, 45523+46035+46547+47059+, 45524+46036+46548+47060+, 45525+46037+46549+47061+, 45526+46038+46550+47062+, 45527+46039+46551+47063+, 45528+46040+46552+47064+, 45529+46041+46553+47065+, 45530+46042+46554+47066+, 45531+46043+46555+47067+, 45532+46044+46556+47068+, 45533+46045+46557+47069+, 45534+46046+46558+47070+, 45535+46047+46559+47071+, 45536+46048+46560+47072+, 45537+46049+46561+47073+, 45538+46050+46562+47074+, 45539+46051+46563+47075+, 45540+46052+46564+47076+, 45541+46053+46565+47077+, 45542+46054+46566+47078+, 45543+46055+46567+47079+, 45544+46056+46568+47080+, 45545+46057+46569+47081+, 45546+46058+46570+47082+, 45547+46059+46571+47083+, 45548+46060+46572+47084+, 45549+46061+46573+47085+, 45550+46062+46574+47086+, 45551+46063+46575+47087+, 45552+46064+46576+47088+, 45553+46065+46577+47089+, 45554+46066+46578+47090+, 45555+46067+46579+47091+, 45556+46068+46580+47092+, 45557+46069+46581+47093+, 45558+46070+46582+47094+, 45559+46071+46583+47095+, 45560+46072+46584+47096+, 45561+46073+46585+47097+, 45562+46074+46586+47098+, 45563+46075+46587+47099+, 45564+46076+46588+47100+, 45565+46077+46589+47101+, 45566+46078+46590+47102+, 45567+46079+46591+47103+,')
```

**Verify other tools disabled/enabled**
Play with the new region selection button (next to the lineplots):

* Activating it should enable the lineplots on the image axes.
* Deactivating it should leave the standard lineplots activated.
* Deactivating the standard lineplots button while region selection is active should deactivate both.
* When region selection is active pan/zoom should be deactivated/disabled. Zoom with the scroll should still work.
* For MD workspaces enabling non-orthogonal axes mode should disable region selection and standard line plots

**ROI selection**
On both matrix and MD workspaces:

* Activate the region selection tool
* Draw a selection
* Releasing the mouse should result in a preview plot in the X/Y direction
* The status bar displays the key presses that export workspaces from the slice viewer
* Press `r` and see that a new workspace is created, suffixed with `_roi`, while the status bar temporarily shows a message as to what has happened. Check the exported region matches your expectations. They can just be reopened in a separate sliceviewer.
* Try transposing and ensuring the exported workspace looks correct.

**1D Cuts***
* From a selected region the status bar shows keys `c`, `x` & `y` will export 1D cuts to workspaces where `c` stands for `c` and is a shortcut to export both X/Y cuts.
* Try each of the keys and check the output workspaces.

1D MD workspaces cannot yet be viewed from a click menu but can be plotted with:
```
from mantid.api import AnalysisDataService
xcut = AnalysisDataService['mdws_cut_x']
fig, ax1 = plt.subplots(subplot_kw={'projection':'mantid'})
ax1.plot(xcut)
plot.show()
```

The cut itself uses BinMD to compute the correct binning whereas the preview approximates the sum from the image so some differences are expected. 

Fixes #27467
Fixes #28966
Fixes #29005 

*This does not require release notes* because **a separate task will update these once final changes to the sliceviewer are complete**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by martyngigg** [comment]
@gvardany Would you have time to review this soon?

**Comment by gvardany** [comment]
@martyngigg Will aim for this week.

**Comment by martyngigg** [comment]
@gvardany Thanks!

**Comment by gvardany** [comment]
Tool button behaviour is as expected.
Region and selection for MD workspace seems fine as well.
However I have some problems with MatrixWorkspace:
- c, x, y don't work for raw data (i.e. in spectrum number) - you can try this by just loading the 317369 in unit test data under ILL/D17.
- for the same file r cuts in y, but not in x
- while working with the slicing it was often crashing with trace leading to `MantidQt::MantidWidgets::WorkspaceTreeWidgetSimple::sliceViewerClicked(QStringList const&) + 55`. I will try to log some steps to reproduce this.
- load the same file, convert the spectrum axis to theta, open it on the resulted workspace, here x and y work, but c produces only x
- load the attached processed file (rename to .nxs), there it does not open up at all, it throws an unknown exception. btw colorfill also throws on this, but it was working recently. These might be related to the numeric axis or units of the axes.
- a general comment on the region selection; when it is too narrow it is no longer possible to grab the side midpoint in order to stretch, it behaves as though holding the centre point, so it moves the selection but does not expand it.
- I have few layout problems (possibly on mac only) that are mostly related to the point readout table (see the screenshots)
- finally, the x, y and c shortcuts should work also with the line plots, is this supposed to be in this PR?

[d16_colorfill.nxs.zip](https://github.com/mantidproject/mantid/files/4937690/d16_colorfill.nxs.zip)

slice-1
slice-2

**Comment by gvardany** [comment]
Also sometimes c would say Invalid value for property `InputWorkspace`; it looks like the workspace name is mixed up somewhere.

**Comment by martyngigg** [comment]
Thank you for testing this out thoroughly.

I had opened a subsequent PR, #28977, when Spencer found a file that threw unexpected exception errors. I believe this makes the SV much more stable with MatrixWorkspace types. Maybe I should just merge that PR here...

**Comment by gvardany** [comment]
Yes, maybe merging would make sense in this case.

**Comment by martyngigg** [comment]
I've rebased the other branch here. This should hopefully fix the errors/crashes.

I'll look into the layout on mac and the cut/slice behaviour that doesn't look quite right next week.

**Comment by gvardany** [comment]
There is another problem with raw data, pointing cursor on the monitor throws an error: two theta is not defined. This could also be better handled. You can try again with 317369.

**Comment by martyngigg** [comment]
I think I have now sorted the issues here; most were due to bad handling of ragged workspaces. When a cut from a ragged workspace is selected the workspace is first internally rebinned to a common grid and the cut/roi then extracted from this.

I have added the cuts from the single-pixel line plots also.

**Comment by gvardany** [comment]
Thanks for the update, now I get the following:

- it starts up as full screen on mac, which is probably not nice
- the issue with the mid-side-point resize is still there, if the slice is too narrow it is no longer possible to stretch from sides. It looks like the active area attached to the center point is too large.
- it stopped working for MD, for the example above, clicking r produces an error: bool is not subscriptable
- for the raw data (see ILL/D17/397827 for example) I get this error on mouse hover: 
```Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/matplotlib/cbook/__init__.py", line 216, in process
    func(*args, **kwargs)
  File "/Users/vardanyan/mantid/qt/python/mantidqt/widgets/sliceviewer/cursortracker.py", line 50, in on_mouse_leave
    self.on_cursor_outside_axes()
  File "/Users/vardanyan/mantid/qt/python/mantidqt/widgets/sliceviewer/imageinfowidget.py", line 45, in on_cursor_outside_axes
    self._widget.cursorAt(DBLMAX, DBLMAX, DBLMAX)
RuntimeError: wrapped C/C++ object of type ImageInfoWidget has been deleted```

- again for 397827, r still cuts in y, but not in x
- for 397827 converted to 2theta (so numeric axis), c produces only x, but not y
- also shall we make the line plot scale to match the colorbar scale in this PR?

**Comment by gvardany** [comment]
Also the labels for the axes are not visible (on mac), can we set a minimum width for the buttons?

Screenshot 2020-07-28 at 13 30 51

**Comment by martyngigg** [comment]
> Also the labels for the axes are not visible (on mac), can we set a minimum width for the buttons?
> 
> Screenshot 2020-07-28 at 13 30 51

That's what I thought I'd figured out. I'll take another look.

**Comment by martyngigg** [comment]
> Thanks for the update, now I get the following:
> 
>     * it starts up as full screen on mac, which is probably not nice

I'm not quite sure how that has happened. Was it full screen last time you closed it and macOS has remembered this?
 
>     * the issue with the mid-side-point resize is still there, if the slice is too narrow it is no longer possible to stretch from sides. It looks like the active area attached to the center point is too large.

I think this is standard behaviour of the matplotlib widget, you have to use the corners to the resize it when it gets very small. Is it too annoying to leave like that?

> 
>     * it stopped working for MD, for the example above, clicking r produces an error: bool is not subscriptable
> 
>     * for the raw data (see ILL/D17/397827 for example) I get this error on mouse hover:
> 
> 
> ```
>   File "/usr/local/lib/python3.7/site-packages/matplotlib/cbook/__init__.py", line 216, in process
>     func(*args, **kwargs)
>   File "/Users/vardanyan/mantid/qt/python/mantidqt/widgets/sliceviewer/cursortracker.py", line 50, in on_mouse_leave
>     self.on_cursor_outside_axes()
>   File "/Users/vardanyan/mantid/qt/python/mantidqt/widgets/sliceviewer/imageinfowidget.py", line 45, in on_cursor_outside_axes
>     self._widget.cursorAt(DBLMAX, DBLMAX, DBLMAX)
> RuntimeError: wrapped C/C++ object of type ImageInfoWidget has been deleted
> ```
> 
>     * again for 397827, r still cuts in y, but not in x
> 
>     * for 397827 converted to 2theta (so numeric axis), c produces only x, but not y
> 
>     * also shall we make the line plot scale to match the colorbar scale in this PR?

Do you mean the scale type,  i.e. linear/log or the min/max? One issue with setting the maximum of the line plot to match the colorbar scale for the rectangle selection is that as it's summing over the other dimension the maximum of the curve is then a lot larger than the max on the scale and you can't see the curve. 

>     * for reduced data (see d16_colorfill above), x,y and c do not work on line plots, but they do with slices.

**Comment by gvardany** [comment]
> I'm not quite sure how that has happened. Was it full screen last time you closed it and macOS has remembered this?

Yes, apparently it is back to normal now.

> I think this is standard behaviour of the matplotlib widget, you have to use the corners to the resize it when it gets very small. Is it too annoying to leave like that?

That's not really critical.

> Do you mean the scale type, i.e. linear/log?

Yes.

**Comment by martyngigg** [comment]
Thanks for the review. I've just pushed some further fixes. Mostly more ragged workspace handling.
I couldn't reproduce the exception with the ImageInfoWidget. I've rebased with master and that has brought in the scale fix for the lineplots to match the colorbar.
This will need a recompile and that might fix the exception. Either that or it is a mac oddity.

File: qt/widgets/common/src/ImageInfoModel.cpp [hunk 1]
```diff
@@ -8,32 +8,23 @@
 #include "MantidQtWidgets/Common/ImageInfoModel.h"
 #include "MantidQtWidgets/Common/QStringUtils.h"
 
-#include <QRegExp>
-#include <iomanip>
 #include <sstream>
 
+namespace {}
```
Comment (gvardany) [inline qt/widgets/common/src/ImageInfoModel.cpp:13]: This seems to be a leftover.

**Review by gvardany — COMMENTED** [review]
Thanks, it is much better now and is fine for MD workspaces. 
I am still unsure if the line plot lin/log scale matches the colorbar; when I plot the cuts in log scale they look very different as compared to what is shown in the line plots.
Two problems persist, possibly related to the ragged workspaces:
- when selecting a slice, `y` is not cut in y:
```
in5 = Load('ILL/IN5/189171')
ws = SofTwoThetaTOF(in5)
```
- in the following case, c does not produce the y (there seems to be an error in the status line, but not in the logs), and x is somehow a 2D workspace:
```
ref = Load('ILL/D17/317370')
ws = ConvertSpectrumAxis(ref, Target='Theta')
```

**Comment by martyngigg** [comment]
I hope  I have fixed these issues now.

**Review by gvardany — APPROVED** [review]
Great work, this works in all the scenarios now. Test coverage is adequate, the release notes will be added afterwards. If I could just add to the wishlist for future; it would be of great use if the slice viewer works on the group of matrixworkspaces in the same way as for MD; that is with the slider through the workspace group index. Would it be easy to implement? Of course this is beyond this PR.

**Comment by martyngigg** [comment]
@gvardany Thanks for reviewing this so thoroughly.

On the group behaviour, just so I understand, do you mean effectively treat the workspace group index as a 3rd dimension? If so then this might be a little fiddly but should be doable. It would probably make sense for it to work for a group of MD workspaces too.

**Comment by gvardany** [comment]
Yes, that's exactly what I meant. Maybe we can open an issue for that.

**Comment by martyngigg** [comment]
@gvardany Sounds good.

**Pull request merged**